### PR TITLE
Support editing container metadata and templates over API

### LINKED ISSUE #2823: Allow for modification of metadata.yaml and templates over API
As I mentioned in #/2821, we should allow for modification of metadata.yaml and templates through the API at the container level. This will allow users to view/edit/add templates and other metadata to a container prior to copying it or publishing it as an image.

The way I see this being done is through:
 - /1.0/containers/NAME/metadata (GET, PATCH, PUT)
 - /1.0/containers/NAME/metadata/templates (GET, POST)
 - /1.0/containers/NAME/metadata/templates/FILENAME (DELETE, GET, PATCH, POST, PUT)

At the CLI level, I'd just add:
 - lxc config metadata show CONTAINER
 - lxc config metadata edit CONTAINER

### PR DESCRIPTION
**Pull request by albertodonato**
This adds API support to get/edit the container metadata.yaml, as well as editing template files.
Also, it adds `lxc config metadata [show|edit] <container>`

Closes #2823

### CODE REVIEW

**Comment by stgraber** [comment]
This needs rebasing and golint is also a bit unhappy about the client/ change.

**Comment by albertodonato** [comment]
@stgraber rebased and updated

**Comment by stgraber** [comment]
jenkins: test this please

File: client/interfaces.go [hunk 1]
```diff
@@ -86,6 +86,9 @@ type ContainerServer interface {
 	GetContainerLogfile(name string, filename string) (content io.ReadCloser, err error)
 	DeleteContainerLogfile(name string, filename string) (err error)
 
+	GetContainerMetadata(name string) (*api.ImageMetadata, string, error)
+	SetContainerMetadata(name string, ETag string, metadata api.ImageMetadata) error
```
Comment (stgraber) [inline client/interfaces.go]: ETag should always be last in the list of arguments.

File: shared/api/image.go [hunk 2]
```diff
@@ -102,3 +102,19 @@ type ImageAliasesEntry struct {
...
+// ImageMetadata represents LXD image metadata
+type ImageMetadata struct {
+	Architecture string                    `json:"architecture" yaml:"architecture"`
+	CreationDate int64                     `json:"creation_date" yaml:"creation_date"`
+	ExpiryDate   int64                     `json:"expiry_date" yaml:"expiry_date"`
+	Properties   map[string]string         `json:"properties" yaml:"properties"`
+	Templates    map[string]*templateEntry `json:"templates" yaml:"templates"`
+}
+
+type templateEntry struct {
```
Comment (stgraber) [inline shared/api/image.go]: Should be renamed to ImageMetadataEntry

File: client/interfaces.go (hunk 1 above)
Comment (albertodonato) [inline client/interfaces.go]: done

File: shared/api/image.go (hunk 2 above)
Comment (albertodonato) [inline shared/api/image.go]: done

File: doc/rest-api.md [hunk 3]
```diff
@@ -1177,6 +1179,111 @@ Return:
 * Operation: Sync
 * Return: empty response or standard error
 
+## /1.0/containers/\<name\>/metadata
```
Comment (stgraber) [inline doc/rest-api.md]: All of the entries below should have:
```
 * Introduced: with API extension "container_edit_metadata"
```

File: lxc/config.go [hunk 4]
```diff
@@ -79,6 +103,11 @@ lxc config show [<remote>:][container] [--expanded]
 lxc config edit [<remote>:][container]
     Edit configuration, either by launching external editor or reading STDIN.
 
+*Container metadata*
+
+lxc config metadata show [<remote>:][container]
+    Show the container metadata.yaml content.
```
Comment (stgraber) [inline lxc/config.go]: Should "lxc config metadata edit" also be listed here?

File: test/main.sh [hunk 5]
```diff
@@ -616,6 +616,12 @@ run_test test_snap_restore "snapshot restores"
 run_test test_config_profiles "profiles and configuration"
 run_test test_config_edit "container configuration edit"
 run_test test_config_edit_container_snapshot_pool_config "container and snapshot volume configuration edit"
+run_test test_metadata_show "container metadata show"
+run_test test_metadata_edit "contaimer metadata edit"
+run_test test_metadata_template_list "contaimer template list"
+run_test test_metadata_template_content "contaimer template content"
+run_test test_metadata_template_add "container template add"
+run_test test_metadata_template_remove "container template remove"
```
Comment (stgraber) [inline test/main.sh]: This ought to be a single entry. The entry can call the sub-functions though.

The problem with adding a lot of "run_test" is that it creates as many sections on Jenkins and makes it a bit harder for me to go through things.

Comment (albertodonato) [inline test/main.sh]: Why is it harder? I'd think it's quite the opposite, tests like basic_usage that do a lot of different things are harder to debug when they fail

File: lxc/config.go (hunk 4 above)
Comment (albertodonato) [inline lxc/config.go]: Done

File: test/main.sh (hunk 5 above)
Comment (stgraber) [inline test/main.sh]: https://jenkins.linuxcontainers.org/view/LXD/view/LXD%20builds/job/lxd-github-pull-test/2763/arch=amd64,backend=dir,compiler=golang-1.6,restrict=lxd-priv/consoleFull will show you the problem. The list of sections on the left already takes several seconds to get extracted and rendered. Gets worse every time we add more stuff.

I agree that "basic_usage" includes way too much stuff including a bunch of duplication with other tests.

What I'd like to see is one "run_test" per feature, so a "test_container_metadata" one is fine, but having things split further just makes things harder, both for dealing with jenkins and if manually running tests (as there's no way I'll remember test_metadata_template_remove so will have to always look at test/main.sh...).

Comment (albertodonato) [inline test/main.sh]: Ok, changed this too

File: doc/rest-api.md (hunk 3 above)
Comment (stgraber) [inline doc/rest-api.md]: @albertodonato did you miss this comment?

I think it's the only thing that still needs fixing before I can merge this.

Comment (albertodonato) [inline doc/rest-api.md]: Sorry, I did miss it. I fixed it now.

File: doc/rest-api.md [hunk 6]
```diff
@@ -1177,6 +1179,117 @@ Return:
...
+### GET (?path=\<template\>)
+* Description: Content of a container template
+* Introduced: with API extension "container\_edit\_metadata"
+* Authentication: trusted
+* Operation: Sync
+* Return: the content of the template
+
+## POST (?path=\<template\>)
+* Description: Add a continer template
+* Authentication: trusted
```
Comment (stgraber) [inline doc/rest-api.md]: you forgot this one :)

**Pull request merged**